### ISSUE TITLE
Setting a PyScript Collective mission, vision, purpose

### ISSUE THREAD

**Issue by pauleveritt** [pyscript/pyscript-collective#22]
### Checklist

- [X] I added a descriptive title
- [X] I searched for other issues and couldn't find a duplication

### What is the issue/comment/problem?

Mission and Vision of the PyScript Collective

_tl;dr Who it is for, why it matters, what it does_

# PyScript Collective: Working together on PyScript resources for learners and contributors

The PyScript Collective organizes community-contributed examples, articles, and other resources to make learning PyScript easy. But that’s not the main thing! The Collective is a group of people that help each other and as well as new contributors, thus getting more folks involved in PyScript and Python.

## Helping Learners and Contributors

We want PyScript to be a great onboarding vehicle for Python, data science, and web coding. Fortunately a lot of folks want to help too! The Collective is a community project to organize everything about PyScript (and related projects like Pyodide) into useful collections.

1) Awesome PyScript is an organized, curated, and updated collection of every…damn…thing. If you have an example, put it up somewhere and let us know. We’ll list it (terms and conditions apply.) Same for an article, a video, a conference talk, whatever.

  We’ll list it, but we won’t stand behind it. We’ll remove it if we’re told it’s no longer good, but that’s it. Not much of a “stamp of approval."

  If you’re looking for a collection of quality, working, maintained stuff, then that would be…

2) The Collective is a shared-responsibility repo of resources, such as examples. As such, it is a group of maintainers who help stuff come in, help clean it up, help teach people, then stand behind what goes in. They’ll fix bugs, refactor examples if PyScript changes, answer tickets, etc. Even better, they’ll help contributors join the Collective and become maintainers. And those new contributors can then help the next person.

  We want the Collective to become a joyful place where people are valued and valuable. Where new coders become the next generation of stars. Hell, we want people to list it on their resume, where the Collective helps them land a job.

  As such, the Collective has a higher bar. It isn’t fair to the maintainers for new contributors to do a “dump and run.” We’ll require: tests, docstrings, and perhaps some code quality tools (flake8, black.) Why? To make it easy to maintain, but also to be shining code examples for learners.

3) The “Incubator” is the in-between stage. Stuff that’s on its way in, isn’t ready to get a permanent commitment yet. Perhaps it’s the code: needs cleaning up, doesn’t use the right idioms. Perhaps its the contributor: they aren’t sure yet if they are around for the long-haul and the maintainers don’t feel comfortable holding the bag.

  Stuff in the “incubator” might not be listed publicly. Such as...

4) The Gallery is the work product of the Collective. Maybe it’s a pile of Sphinx docs that get linked into the main PyScript docs. Maybe it is a standalone website with a catalog-style navigation.

  Maybe…just maybe…it’s an installable PyScript PWA where you can run the examples and (gasp) edit little hacks in the browser, to experiment.

## Will This Work?

It’s kind of ambitious, in a social way. Will we get maintainers? Will people want to join? We think so. Rather, we very-much *hope* so. We’d like PyScript to be a vehicle for crazy-cool innovation, growth, and togetherness, similar to how data science impacted Python.

## Kinda Vague, What Specifically?

- We’ll get a bootstrapping set of maintainers who commit to building the machinery (easy testing, GitHub Actions, flake8 etc., Sphinx output, Awesome PyScript Markdown page) and closing tickets

- Collect all the existing known resources into the Awesome PyScript page

- Make it clear the process to contribute to Awesome PyScript

- Start processing our own examples through the process to see how well it works

- Make it clear the process to contribute to the Collective

- See if anybody wants their examples in the Collective, perhaps to join us

**Comment by pauleveritt** [comment]
Hi, this is the direction that we've talked with a few people about. @tildebyte @marimeireles John, Tanaka, Fabio, and Luciano have taken a look at it. Somewhat mirrored on the "Plone Collective" that I was loosely involved with back-in-the-day.

Specifically interested in who wants to join and help others through co-ownership of examples.

**Comment by tildebyte** [comment]
> 1. Awesome PyCharm

is this a typo?

Sheer personal opinion on my part: please don't use `black`.

`flake8` can do everything `black` can do (with the appropriate config; and has pretty sane defaults), but allows the Collective to decide what they want[1].

`flake8` also has a bajillion plugins for everything from code complexity to naming conventions to security - not that the Collective **necessarily** needs all of those things **right now** - but there are some (naming convention, and some plugins for other kinds of conventions _including e.g pytest_) which *would* be nice to have.

I'll have more to say about the actual content herein after I mull it over a bit more 😁

**Comment by pauleveritt** [comment]
Apologies, totally a typo on the Awesome PyCharm part. I will edit it.

If Black becomes a bikeshed magnet then we'll drop it. But of course, almost any code style decision becomes a bikeshed. Which ironically was the motivation of Black. Everybody sacrifice their sovereignty because some decision is better than none.

**Comment by tildebyte** [comment]
Too true, too true :)

**Comment by tildebyte** [comment]
To make sure that **I** understand:

> 1. Awesome PyScript

This is a single file, in **this** repo, which is a list of user-submitted links (and presumably at least minimally vetted by a human or some process)[1]

> 2. The Collective is a shared-responsibility repo of resources

This repo in toto

> 3. The “Incubator” is the in-between stage

PRs in this repo? A sub-repo? A separate folder for WIPs?

> 4. The Gallery

A showcase webpage of fully-reviewed, merged PRs which presumably is not unlike https://threejs.org/examples (these are "official", maintained examples whose source lives in the main repo). The Collective "owns" these (at minimum, in the sense of being responsible for their care and feeding. c.f. _felis catus_)

[1] I'd love it if we had something like https://threejs.org <- That's the actual three.js homepage, with a T.O.C. for project links on the left (10%) - the rest (other 90%) is a gallery of links( with thumbnails!) submitted by users.

**Comment by pauleveritt** [comment]
1. Correct.
2. Correct. Most importantly, the people here.
3. TBD. Maybe just PRs...maybe a parallel `examples`.
4. Correct. Could be anything, I bet we have a blank slate. (Though the threes.org home page didn't load in Safari, content blocker.)

I'm most interested to find out if you, @jezdez @ntoll @jhanley-com and Tanaka are up for the hard work as bootstrappers.

**Comment by ntoll** [comment]
Yes. Would love to be involved. But currently up to my eyes at EuroPython. :smile:

**Comment by tildebyte** [comment]
I am interested and motivated, but unfortunately I don't have the ability to make a commitment to _regular_ contributions - sometimes I have a whole weekend free, sometimes I have *no* free time for several weeks 😦 🤷  and (of course 🙄) it's not predictable...

My office hours can be "expect me when you see me" 😆

**Comment by pauleveritt** [comment]
@ntoll I suspect you will have a fairly monumental PyScript contribution in several regards. 😀 It's no rush at all, and the most important thing you can give us is wisdom from your last ten years doing something similar.

@tildebyte That's fine. It's my job to kind of be the daily glue, aggragating effort from folks who come-and-go. If that what you can give, that's fine. I'm more interested in whether you believe in this idea (let's go make people awesome.)

**Comment by tildebyte** [comment]
> whether you believe in this idea (let's go make people awesome)

ABSOLUTELY! 😁  The best way (IMO and experience) for new projects to gain traction is to have a smooth onramp, including LOTS of detailed example code which does interesting and meaningful things at all levels of experience.

Counterexample: I'm currently trying to work with an API which is new to me (not naming names 😬), which is amazing in many ways, **except** - their doc site search is so awful, I have to switch out to DDG, and their examples essentially go from "Hello, World" to "Build Entire World" with *nothing* in-between...

**Comment by jhanley-com** [comment]
Yes, I am available to be a bootstrapper.

**Comment by pauleveritt** [comment]
@marimeireles That's good to know about the resource compendium. I guess this Collective gallery will just be examples then and act as input to the other thing.

The bare-bones Awesome PyScript Markdown collection though can still list everything (articles etc.), right?

On your first point...I'm about to start vacation tomorrow. I'll load this into my head and try to make some kind of Collective site prototype for us to work through.

**Comment by pauleveritt** [comment]
As update, I've spent my vacation working on collective stuff, which I hope to share in a couple of days. It's a lot of interconnected parts. For now, I'll add as repo collaborators: @jhanley-com @tildebyte 

Let's see if we can get the PRs worked through and incoming tickets.

**Comment by pauleveritt** [comment]
Gave write to both.

**Comment by fpliger** [comment]
Thank you @pauleveritt for leading the effort and to everyone wanting to help. This is so exciting!

**Comment by marimeireles** [comment]
Thanks @pauleveritt! Great work! :D

**Comment by pauleveritt** [comment]
Hi to @tildebyte @jhanley-com @ntoll  I’m at a checkpoint on my vacation project. 

A prototype of a PyScript Collective site:

- A Gallery web app and static-served website
- Runnable locally and offline via pipx or python -m
- A new approach to PyScript example testing

Here’s a video walkthrough:

  https://youtu.be/a9MDiAMf7Kw    

Here’s the prototype hosted site:

  https://pauleveritt.github.io/psc_io/

And here is a diary I wrote while making this:

  https://pscdemo.readthedocs.io/en/latest/building/index.html

For example, notes on how to do Playwright testing without an HTTP server:

  https://pscdemo.readthedocs.io/en/latest/building/playwright.html

**Comment by tildebyte** [comment]
> it’s written as if it is a Python package, meant to be installed

This is seriously genius

**Comment by pauleveritt** [comment]
Thanks @tildebyte 

If there was interest in this kind of direction, I think I might re-start on the diary. I can of let-go-of-the-rope at the end, ran out of time.

This time, I'd like to go step-at-a-time and (a) make it into _our_ implementation (so, discussion) and (b) teach everybody how it was done.

Would that be worth it?

**Comment by tildebyte** [comment]
Yes 😃 

I'm reading through the diary now - it's very informative, but I'm not *that* familiar with web testing per-se, so I'm sort of white-knuckling it 😁

**Comment by tildebyte** [comment]
Kudos on going full-Python (and Markdown!), rather than some awful webby/javascripty chimera (npm ugh)

Some notes, ordered from most rational to most emotional 😆

- codecov: Setting up codecov a few years ago for my day job was so broken and awful, we
  decided to go with coveralls: we have no regrets
- pycharm: I have no objection to notes for PyCharm users, but I'd like to make
  sure that we dont inadvertently e.g. make it hard for NON-PyCharm users to
  follow tutorials, etc.
- click: It's not my choice for CLIs, but it's perfectly adequate
- black: As previously mentioned... really, it's some of the style choices
  which I strongly disagree with, not the idea of black itself (have you heard of [blue][] 😝?)

[blue]: https://blue.readthedocs.io/en/latest/

**Comment by pauleveritt** [comment]
Well, all the other places can your jam. A nice web app, etc.

FTR, the part about building artifacts for static export...that was certainly race-to-the-finish. Needs starting over.

**Comment by pauleveritt** [comment]
If the price of getting you to be co-pilot is black->blue, then count me in. At least maybe. 😀 Need to ensure it uses the same `blackd` that I can point my IDE at. And other tooling e.g. pre-commit.

I don't mind switching from typer/click to something else, as long as YOU do it. 😈 I know very little about that.

FTR, something I care a lot about, for future...the web app has a really cool reloader. Something that is competitive with what the frontends do.

**Comment by tildebyte** [comment]
I'm cool with click and typer!

I'm also cool with black! It *is* actually configurable to a degree, and the (perhaps surprising) thing I care most about can be ignored with a simple `--skip-string-normalization`[1]

My only other comment about black would be to actually *follow* their advice about line length:

(cherry-picking here)
> In general, 90-ish seems like the wise choice
...
> it’s better if you use [Bugbear](https://github.com/PyCQA/flake8-bugbear)’s {a flake8 plugin} B950 warning instead of {flake8's} E501, and bump the max line length to ... the --line-length you used for black, which will align more with black’s “try to respect --line-length, but don’t become crazy if you can’t”

TL;DR:  Use Bugbear's B950, and set it and black's line-length to something between 88 and 92

[1]  *Personally*, I prefer `'` for everything possible - seeing lots and lots of `"` starts to look like line noise, particularly in monospaced fonts, particularly at the small font sizes I tend to use. flake8 will still catch outright erroneous use of quotes.

**Comment by tildebyte** [comment]
> Well, all the other places can your jam. A nice web app, etc.

Sorry, it took me a minute to get this... I think? Coveralls has a nice, reliable (cough codecov cough) webapp as well...

**Comment by pauleveritt** [comment]
> > Well, all the other places can your jam. A nice web app, etc.
> 
> Sorry, it took me a minute to get this... I think? Coveralls has a nice, reliable (cough codecov cough) webapp as well...

Apologies, vacation typing. 😇 If other parts are what interests you, then focus on that. Most of all, help me steer this ship into port.

**Comment by pauleveritt** [comment]
> I'm cool with click and typer!
> 
> I'm also cool with black! It _is_ actually configurable to a degree, and the (perhaps surprising) thing I care most about can be ignored with a simple `--skip-string-normalization`[1]

I propose we move this to my `psc` repo. Make some PRs on any tooling change/re-configuration you'd like. I'm still learning, and would enjoy the conversation.

**Comment by pauleveritt** [comment]
@jhanley-com the checkpoint above from 5 days ago...is this something you'd be interested in working on?

**Comment by pauleveritt** [comment]
I'm going to close this ticket, we'll get started on the next ones.
